components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +73,158 @@ impl MutationObserver {
         ScriptThread::add_mutation_observer(&*observer);
         Ok(observer)
     }
+
+    /// https://dom.spec.whatwg.org/#queue-a-mutation-observer-compound-microtask
+    /// Queue a Mutation Observer compound Microtask.
+    pub fn queueMutationObserverCompoundMicrotask() {
+        // Step 1
+        if ScriptThread::is_mutation_observer_compound_microtask_queued() {
+            return;
+        }
+        // Step 2
+        ScriptThread::set_mutation_observer_compound_microtask_queued(true);
+        // Step 3
+        let compoundMicrotask = Microtask::NotifyMutationObservers();
+        ScriptThread::enqueue_microtask(compoundMicrotask);
+    }
+
+    /// https://dom.spec.whatwg.org/#notify-mutation-observers
+    /// Notify Mutation Observers.
+    pub fn notifyMutationObservers() -> Fallible<()> {
+        // Step 1
+        ScriptThread::set_mutation_observer_compound_microtask_queued(false);
+        // Step 2
+        let notifyList = ScriptThread::get_mutation_observer();
+        // Step 3, Step 4 not needed as Servo doesn't implement anything related to slots yet.
+        // Step 5: Ignore the specific text about execute a compound microtask.
+        for (i, mo) in notifyList.iter().enumerate() {
+            let queue: Vec<Root<MutationRecord>> = mo.record_queue.borrow().clone();
+            *mo.record_queue.borrow_mut() = Vec::new();
+            // TODO: Step 5.3 Remove all transient registered observers whose observer is mo.
+            if !queue.is_empty() {
+                let callback: MutationCallback = *mo.callback;
+                callback.Call_(&*mo, queue, &*mo, ExceptionHandling::Report);
+            }
+        }
+        // Step 6 not needed as Servo doesn't implement anything related to slots yet.
+        Ok(())
+    }
+
+    //https://dom.spec.whatwg.org/#queueing-a-mutation-record
+    //Queuing a mutation record
+    pub fn queue_a_mutation_record(target: &Node, attr_type: Mutation) {
+        // Step 1
+        let mut interestedObservers: Vec<Root<MutationObserver>> = vec![];
+        let mut pairedStrings: Vec<DOMString> = vec![];
```

Comment:
We can do `let mut interesedObservers: Vec<(Root<MutationObserver>, DOMString)> = vec![]` rather than having two separate vectors. This will avoid complications with matching up indexes.